README.md
# aristaZTP
My version of Arista ZTP using CSV file

# Assumptions: I installed on ubuntu 16.04 and run the setup script

The CSV file needs to be created and available in the tftpboot directory

The template.txt file need to be in the tftpboot directory

When the ZTP switch boot it looks for the file to download from the dhcp settings.

In this case boot.py is downloaded to the switch and executed.

The CSV and Template are copied over and the CSV file is read

A process loop looks through the CSV file to find a MAC address match for the swicth.

The Template function is used to replace the variables in the template.txt file and
it is saved as the startup-config.

Script will use the 172.2.0.10-172.20.0.250 address space for the DHCP server.
This can be adjusted by editing the script file.

The switch reboots.

You are now ready to manage the switch with my Ansible repo here:

https://github.com/xod442/eos-ansible-quick-start.git

Or add the switches into the Cloud Vision Portal Server LQQK Here!:

http://techworldwookie.com/?p=205

You can maintain the CSV file with switchdb located here:

https://github.com/xod442/scriptsonly/tree/master/switchdb


CHANGES.md
# CHANGES.md

version 0.1.2
- added the CHANGES.md and updated the README.md file


setup.sh
# Arista ZTP CSV server configuration script. 
#
#      When finished..... 'sudo chmod 777 /var/lib/dhcp/dhcpd.leases'
#      To check dhcp run 'sudo dhcp -f' look for errors
#--------------------------------------------------------------------------------
#
#       Update & Upgrade, Uncomment if necessary
#
#--------------------------------------------------------------------------------
#chmod 777 ./setup.sh
#apt-get update
#apt-get -y upgrade
#apt-get -y install git
#apt-get -y install python-pip
#apt -y autoremove
#--------------------------------------------------------------------------------
#
#       Install dhcp server and configure conf file
#
#--------------------------------------------------------------------------------
#apt-get -y install isc-dhcp-server
cp /etc/dhcp/dhcpd.conf /etc/dhcp/dhcpd_bak.conf
echo 'default-lease-time 3600;' > /etc/dhcp/dhcpd.conf
echo 'authoritative;' >> /etc/dhcp/dhcpd.conf
echo 'max-lease-time 3600;' >> /etc/dhcp/dhcpd.conf
echo 'ddns-update-style none;' >> /etc/dhcp/dhcpd.conf
echo 'subnet 172.20.0.0 netmask 255.255.255.0 {' >> /etc/dhcp/dhcpd.conf
echo '    range 172.20.0.10 172.20.0.250;' >> /etc/dhcp/dhcpd.conf
echo '    option tftp-server-name "172.20.0.1";' >> /etc/dhcp/dhcpd.conf
echo '    option bootfile-name "boot.py";' >> /etc/dhcp/dhcpd.conf
echo '}' >> /etc/dhcp/dhcpd.conf
service isc-dhcp-server restart
#--------------------------------------------------------------------------------
#
#       Configure local interface. Skip this step if it is already configured
#
#--------------------------------------------------------------------------------
#echo 'source-directory /etc/network/interfaces.d' > /etc/network/interfaces
#echo 'auto lo' >> /etc/network/interfaces
#echo 'iface lo inet loopback' >> /etc/network/interfaces
#echo 'auto ens33' >> /etc/network/interfaces
#echo 'iface ens33 inet static' >> /etc/network/interfaces
#echo '    address 172.20.0.3' >> /etc/network/interfaces
#echo '    netmask 255.255.255.0' >> /etc/network/interfaces
#Edit these for your environment
#ifdown ens33 && ifup ens33
#ifconfig | grep addr
#pip install pyyaml
#echo $PATH
#--------------------------------------------------------------------------------
#
#       Install tftp server and set home directory
#
#--------------------------------------------------------------------------------

#apt-get -y install xinetd tftpd tftp
echo 'service tftp' > /etc/xinetd.d/tftp
echo '{' >> /etc/xinetd.d/tftp
echo 'protocol = udp' >> /etc/xinetd.d/tftp
echo 'port = 69' >> /etc/xinetd.d/tftp
echo 'socket_type = dgram' >> /etc/xinetd.d/tftp
echo 'wait = yes' >> /etc/xinetd.d/tftp
echo 'server = /usr/sbin/in.tftpd' >> /etc/xinetd.d/tftp
echo 'server_args = /home/rick/projects/aristaZTP/tftpboot' >> /etc/xinetd.d/tftp
echo 'disable = no' >> /etc/xinetd.d/tftp
echo 'user = nobody' >> /etc/xinetd.d/tftp
echo '}' >> /etc/xinetd.d/tftp
Edit these to match your tftpboot directory
chmod -R 777 /home/rick/projects/aristaZTP/tftpboot
chown -R nobody /home/rick/projects/aristaZTP/tftpboot
service xinetd restart
#--------------------------------------------------------------------------------
#
#       Install ftp server 
#
#--------------------------------------------------------------------------------
#apt-get -y install vsftpd
cp /etc/vsftpd.conf /etc/vsftpd.conf.bak
sed -i 's/#write_enable=YES/write_enable=YES/' /etc/vsftpd.conf
sed -i 's/#local_umask/local_umask/' /etc/vsftpd.conf
sed -i 's/#chroot_local_user/chroot_local_user/' /etc/vsftpd.conf
echo 'allow_writeable_chroot=YES' >> /etc/vsftpd.conf
echo 'pasv_enable=YES' >> /etc/vsftpd.conf
echo 'pasv_min_port=40000' >> /etc/vsftpd.conf
echo 'pasv_max_port=40100' >> /etc/vsftpd.conf
service vsftpd restart

netstat -plnut4




tftpboot/boot.py
#!/usr/bin/env python

'''
 Copyright 2016 Hewlett Packard Enterprise Development LP.
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at
       http://www.apache.org/licenses/LICENSE-2.0
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
__author__ = "@netmanchris"
__copyright__ = "Copyright 2016, Hewlett Packard Enterprise Development LP."
__credits__ = ["Rick Kauffman"]
__license__ = "Apache2"
__version__ = "1.0.0"
__maintainer__ = "Rick Kauffman"
__email__ = "rick@rickkauffman.com"
__status__ = "Prototype"
'''

# Arista  by CSV  wookieware 2016

# uses the python "string" import to access the Template function

import os
import sys
import csv
from string import Template
import json
import subprocess

# Create application variables
#--------------------------------------------------

count = 0
ver ='FastCli -p 15 -c "show version | incl address"'

copytemplate ='FastCli -p 15 -c "copy tftp://172.20.1.1/template.txt flash:/template.txt"'
copycsv ='FastCli -p 15 -c "copy tftp://172.20.1.1/varMatrix.csv flash:/varMatrix.csv"'

#--------------------------------------------------

def copyfiles():

	os.system(copytemplate)

	os.system(copycsv)


def configureSwitch():

	switchMacAddressInfo = subprocess.Popen([ver],stdout=subprocess.PIPE, shell=True)
        (out, err) = switchMacAddressInfo.communicate()

        # Extract MAC address from output

	switchMacAddress = out[21:95]
	# Set counter to track items
	count = 0

	#open the template
        print 'Reading template file..........'
	form = open('/mnt/flash/template.txt', 'r')
	src = Template( form.read() )


	#open the csv file and build List
        print 'Reading CSV file............'
	csvfile = open('/mnt/flash/varMatrix.csv', 'r')
	content = csv.DictReader(csvfile)
	switch = []
	for row in content:
		switch.append(row)

	# how many switches do we have?
	check = len(switch)

	while (count < check):

		#Substitute CSV variables in the template file
		myMac = switch[count]['mac'].strip()
		switchMacAddress = switchMacAddress.strip()


		if str(switchMacAddress) == str(myMac):
			result = src.substitute(switch[count])
			config = open('/mnt/flash/startup-config','w')
			config.write(result)
		count = count + 1

	# Be nice and close the files

	csvfile.close()
	form.close()

def cleanup():
	# Clean this mess up else we all wind up in jail
	pass

copyfiles()
#
configureSwitch()
cleanup()
quit()


tftpboot/x.py
#!/usr/bin/env python
# Arista  by CSV  wookieware 2016

# uses the python "string" import to access the Template function

import os
import sys
import csv
from string import Template
import json
import subprocess

# Create application variables
#--------------------------------------------------

count = 0
ver ='FastCli -p 15 -c "show version | incl address"'

copytemplate ='FastCli -p 15 -c "copy tftp://172.20.1.1/template.txt flash:/template.txt"'
copycsv ='FastCli -p 15 -c "copy tftp://172.20.1.1/varMatrix.csv flash:/varMatrix.csv"'

#--------------------------------------------------

def copyfiles():

	os.system(copytemplate)

	os.system(copycsv)

	
def configureSwitch():

	switchMacAddressInfo = subprocess.Popen([ver],stdout=subprocess.PIPE, shell=True)
        (out, err) = switchMacAddressInfo.communicate()

        # Extract MAC address from output
    
	switchMacAddress = out[21:95]
	# Set counter to track items
	count = 0

	#open the template
        print 'Reading template file..........'
	form = open('template.txt', 'r')
	src = Template( form.read() )

	
	#open the csv file and build List
        print 'Reading CSV file............'    
	csvfile = open('varMatrix.csv', 'r')
	content = csv.DictReader(csvfile)
	switch = []
	for row in content:
		switch.append(row)

	# how many switches do we have? 
	check = len(switch) 

	while (count < check):
		
		#Substitute CSV variables in the template file
		myMac = switch[count]['mac'].strip()
		switchMacAddress = switchMacAddress.strip() 


		if str(switchMacAddress) == str(myMac):
			result = src.substitute(switch[count])
			config = open('startup-config','w')
			config.write(result)
		count = count + 1

	# Be nice and close the files
	
	csvfile.close()
	form.close()

def cleanup():
	# Clean this mess up else we all wind up in jail
	pass

copyfiles()
#
configureSwitch()
cleanup()
quit()
